src/aiogram_dialog/manager/registry.py

```diff
@@ -139,91 +124,136 @@ def register(
         group = dialog.states_group()
         if group in self.dialogs:
             raise ValueError(f"StatesGroup `{group}` is already used")
-        self.dialogs[group] = dialog
-        self.state_groups[dialog.states_group_name()] = group
-
-        dialog.register(
-            router if router else self.default_router,
-            IntentFilter(aiogd_intent_state_group=group),
-            *args, **kwargs,
+        self.dialogs[group] = _DialogConfig(
+            dialog=dialog,
+            router=router,
+            args=args,
+            kwargs=kwargs,
         )
 
-    def register_start_handler(self, state: State):
+    def find_dialog(self, state: State) -> DialogProtocol:
+        try:
+            return self.dialogs[state.group].dialog
+        except KeyError as e:
+            raise UnregisteredDialogError(
+                f"No dialog found for `{state.group}`"
+                f" (looking by state `{state}`)",
+            ) from e
+
+    def _state_groups(self) -> Dict[str, Type[StatesGroup]]:
+        return {
+            d.dialog.states_group_name(): d.dialog.states_group()
+            for d in self.dialogs.values()
+        }
+
+    def register_start_handler(self, router: Router, state: State):
         async def start_dialog(
                 message: Message, dialog_manager: DialogManager,
         ) -> None:
             await dialog_manager.start(state, mode=StartMode.RESET_STACK)
 
-        self.dp.message.register(
+        router.message.register(
             start_dialog, Command(commands="start"), any_state,
         )
 
-    def _register_middleware(self):
+    def setup_dp(
+            self,
+            dp: Dispatcher,
+            default_router: Optional[Router] = None,
+    ):
+        update_handler = DialogEventObserver(
+            router=dp, event_name=DIALOG_EVENT_NAME
+        )
+        dp.observers[DIALOG_EVENT_NAME] = update_handler
+        self._register_update_handler(
+            handle_update, update_handler,
+        )
+        self._register_middleware(dp, update_handler)
+        self._register_dialogs(dp, default_router)
+
+    def _register_dialogs(self, dp: Dispatcher, default_router: Router):
+        default_router = (
+                default_router or
+                Router(name="aiogram_dialog_router")
+        )
```

Comment:
The decision to utilize an existing router or create a new one should be made within the `setup_dp` method. Alternatively, and a less favorable option, the type hint should be modified to `Optional[Router]`.